File: function_test/test_util.R
# Test harness for UTIL::getWeights - R side
# Compares R's dbeta() against C++ standalone getWeights()

options(digits = 15)

mafs <- c(0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5)

# Beta(1, 25) -- SAIGE default
cat("=== getWeights with Beta(1, 25) ===\n")
for (maf in mafs) {
    w <- dbeta(maf, 1, 25)
    cat(sprintf("getWeights_beta1_25_maf%s: %.15f\n", maf, w))
}

# Beta(1, 1) -- uniform
cat("=== getWeights with Beta(1, 1) ===\n")
for (maf in mafs) {
    w <- dbeta(maf, 1, 1)
    cat(sprintf("getWeights_beta1_1_maf%s: %.15f\n", maf, w))
}

# Linear kernel (always 1)
cat("=== getWeights with linear kernel ===\n")
for (maf in mafs) {
    cat(sprintf("getWeights_linear_maf%s: %.15f\n", maf, 1.0))
}

File: function_test/test_util.cpp
// Test harness for UTIL::getWeights
// Compares C++ standalone dbeta output against R's dbeta

#include <iostream>
#include <iomanip>
#include <armadillo>
#include "UTIL.hpp"

int main() {
    std::cout << std::setprecision(15) << std::fixed;

    // MAF values to test
    std::vector<double> mafs = {0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5};

    // Default beta params: a1=1, a2=25
    arma::vec wBeta_default = {1.0, 25.0};

    std::cout << "=== getWeights with Beta(1, 25) ===" << std::endl;
    for (double maf : mafs) {
        double w = getWeights("linear.weighted", maf, wBeta_default);
        std::cout << "getWeights_beta1_25_maf" << maf << ": " << w << std::endl;
    }

    // Also test with Beta(1, 1) -- uniform
    arma::vec wBeta_uniform = {1.0, 1.0};

    std::cout << "=== getWeights with Beta(1, 1) ===" << std::endl;
    for (double maf : mafs) {
        double w = getWeights("linear.weighted", maf, wBeta_uniform);
        std::cout << "getWeights_beta1_1_maf" << maf << ": " << w << std::endl;
    }

    // Test linear kernel
    std::cout << "=== getWeights with linear kernel ===" << std::endl;
    for (double maf : mafs) {
        double w = getWeights("linear", maf, wBeta_default);
        std::cout << "getWeights_linear_maf" << maf << ": " << w << std::endl;
    }

    return 0;
}
